tests/zfs-tests/tests/functional/fault/auto_spare_ashift.ksh

```diff
@@ -0,0 +1,100 @@
+#!/bin/ksh -p
+
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2017 by Intel Corporation. All rights reserved.
+# Copyright 2017, loli10K <ezomori.nozomu@gmail.com>. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/include/math.shlib
+. $STF_SUITE/tests/functional/fault/fault.cfg
+
+#
+# DESCRIPTION:
+# Testing Fault Management Agent ZED Logic - Automated Auto-Spare Test when
+# drive is faulted and a custom ashift value needs to be provided to replace it.
+#
+# STRATEGY:
+# 1. Create a pool from 512b devices and set "ashift" pool property accordingly
+# 2. Add one 512e spare device (4Kn would generate IO errors on replace)
+# 3. Inject IO errors with a zinject error handler
+# 4. Start a scrub
+# 5. Verify the ZED kicks in the hot spare and expected pool/device status
+# 6. Clear the fault
+# 7. Verify the hot spare is available and expected pool/device status
+#
+
+verify_runnable "both"
+
+function cleanup
+{
+	log_must zinject -c all
+	destroy_pool $TESTPOOL
+	rm -f $SAFE_DEVICE $FAIL_DEVICE
+}
+
+log_assert "ZED should replace a device using the configured ashift property"
+log_onexit cleanup
+
+SAFE_DEVICE="$TEST_BASE_DIR/safe-dev"
+FAIL_DEVICE="$TEST_BASE_DIR/fail-dev"
+
+# 1. Create a pool from 512b devices and set "ashift" pool property accordingly
+for vdev in $SAFE_DEVICE $FAIL_DEVICE; do
+	truncate -s $SPA_MINDEVSIZE $vdev
+done
+log_must zpool create -f $TESTPOOL mirror $SAFE_DEVICE $FAIL_DEVICE
+# NOTE: file VDEVs should be added as 512b devices, verify this "just in case"
+for vdev in $SAFE_DEVICE $FAIL_DEVICE; do
+	verify_eq "9" "$(zdb -e -l $vdev | awk '/ashift: /{print $2}')" "ashift"
+done
+log_must zpool set ashift=9 $TESTPOOL
+
+# 2. Add one 512e spare device (4Kn would generate IO errors on replace)
+# NOTE: must be larger than the existing 512b devices, add 32m of fudge
+if lsmod | grep scsi_debug >/dev/null; then
+   log_must modprobe -r scsi_debug
+fi
+load_scsi_debug $(($SPA_MINDEVSIZE/1024/1024+32)) $SDHOSTS $SDTGTS $SDLUNS '512e'
+SPARE_DEVICE=$(get_debug_device)
```

Comment:
You’re creating a ‘512e’ scsi device here, but you never reload the ‘scsi_debug’ module to recreate the ‘512b’ device. You can’t depend on test ordering.

Since we need different device types, would it be worth considering having each test in this test group perform a ‘load_scsi_debug’ call to set up an appropriate scsi device?

Reply:
@dinatale2 sorry for the late reply, i've been unable to work on this due to an unexpected workload.

Loading the scsi_debug driver in each test group sounds good to me; i'm also trying to update the other test case to exercise ZED with multiple, parallel faults. I'm going to push something once my local builder is able to run these changes, hopefully before next monday.

OT: does it make sense to split this into two separate commits ("update ZED to handle ashift" + "new spare test case with multiple faults")? I don't mind the extra (minimal) step required if it's going to help keep the git history more readable.

Reply:
@loli10K I'm OK with this either as a single commit or as two.  All I ask if that you open a new PR for the multiple fault test case if you do decide to split it.  It makes it easier for me to merge.

Reply:
Then let's merge this in a single commit, it seems easier for everyone.

On a more related note, moving the 'load_scsi_debug' call into the individual scripts that need it enabled some tests that were previously skipped (`scsi_debug` was missing, and triggering a `log_unsupported` in the setup script): now 'auto_spare_002_pos' seems to be failing on a couple of Ubuntu builders (Ubuntu 17.04 x86_64 / Ubuntu 16.04 x86_64).

From http://build.zfsonlinux.org/builders/Ubuntu%2016.04%20x86_64%20%28TEST%29/builds/1486/steps/shell_8/logs/log:

```
13:24:38.47 NOTE: Performing local cleanup via log_onexit (cleanup)
13:24:38.48   pool: testpool
13:24:38.48  state: ONLINE
13:24:38.48 status: One or more devices has experienced an unrecoverable error.  An
13:24:38.48 	attempt was made to correct the error.  Applications are unaffected.
13:24:38.48 action: Determine if the device needs to be replaced, and clear the errors
13:24:38.48 	using 'zpool clear' or replace the device with 'zpool replace'.
13:24:38.48    see: http://zfsonlinux.org/msg/ZFS-8000-9P
13:24:38.48   scan: none requested
13:24:38.48 config:
13:24:38.48 
13:24:38.48 	NAME             STATE     READ WRITE CKSUM
13:24:38.48 	testpool         ONLINE       0     0     0
13:24:38.48 	  mirror-0       ONLINE       0     0     0
13:24:38.48 	    /mnt/file-1  ONLINE       0     0   142
13:24:38.48 	    /mnt/file-2  ONLINE       0     0     0
13:24:38.48 	    /mnt/file-3  ONLINE       0     0     0
13:24:38.48 	    /mnt/file-4  ONLINE       0     0     0
13:24:38.48 	spares
13:24:38.48 	  /mnt/spare-1   AVAIL   
13:24:38.48 
13:24:38.48 errors: No known data errors
```

For some reason ZED doesn't kick in the hot spare. I've been trying to reproduce this for the past few hours on a local builder (Ubuntu 16.04 x86_64) without luck, the test is always a PASS here.